Scalable LSC field split preconditioning

lindsayad opened the pull request:
See also https://gitlab.com/petsc/petsc/-/merge_requests/6642

Refs #24548

lindsayad committed:
Create a problem dedicated to LSC and test field splits

- Allow using FV variables in matrix tag APIs
- Demonstrate creation of a mass matrix
- Add finite element mass matrix test
- Create NavierStokesProblem
- Add const to more matrix tag apis
- Use KSPSetPreSolve
- "Working" LSC with Qv scaling demo'd on lid_driven finite element input
- Early return from computeResidual if there are no vector tags
- Add off-diagonal computation to PressureGradient
- Compute own L in NavierStokesProblem
- With my L, I can low use LU as the lsc_pc_type
- LSC appears to be scaling independent of mesh size
- Add the ability to do Picard iterations
- Temporarily remove split size sum error checking to allow nested field splits
- Demonstrate nested split by variable
- Demonstrate splitting by sides at top level
- Only associate the Qv mass matrix
- Code for doing nested field splits in NS
- Demonstrate nested field splits in NS
- Demonstrate dirichlet splitting
- Error check for LSC preconditioner
- Error check  for field split preconditioners

Refs #24548

lindsayad committed:
Field split research on top of hybrid cg-dg

- SelfAux -> ProjectionAux
- Take NavierStokesProblem from field split research branch
- Add const to more matrix tag apis
- Allow using FV variables in matrix tag APIs
- Demonstrate creation of a mass matrix
- Add finite element mass matrix test
- Apply FSP to hybrid DG-CG
- Create a matrix_only parameter in the tagging interface
- Check-in a great Q2Q1 field split demo
- Rename Mass -> MassKernel to avoid conflict with TM
- Add test for taylor-hood field-split input
- Modifications for 3D
- Checkin large run PBS script

refs #24548

lindsayad commented:
will get back to this soon now that petsc has been updated

lindsayad committed:
Remove lid_driven_fsp.i and associated tests

lindsayad committed:
Move lid_driven_fs.i back to 2D which should match the gold file

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1898897/) on 22c52b6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24883/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24883/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e8ff3c477129ce4489a2991da30acf0c642c0747`

lindsayad committed:
Remove Mass kernel because it conflicts with TM

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1899100/) on 97d9bbd wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24883/site)

This comment will be updated on new commits.

lindsayad committed:
Add default-arg for std::vector param

lindsayad committed:
Revert "Remove lid_driven_fsp.i and associated tests"

This reverts commit e1df290bc2c3e5ba82cc99079cc9e80d3b17c23e.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1937064/) on 087fe87 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>a95084</th>
      <th colspan=4>#24883 087fe8</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.01%</td>
      <td>84.98%</td>
      <td>-0.03%</td>
      <td>50.76%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>97333</td>
      <td>97389</td>
      <td>+56</td>
      <td>67</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17159</td>
      <td>17207</td>
      <td>+48</td>
      <td>65</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24883/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24883/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>a95084</th>
      <th colspan=4>#24883 087fe8</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.22%</td>
      <td>84.03%</td>
      <td>-0.20%</td>
      <td>67.24%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>13189</td>
      <td>13297</td>
      <td>+108</td>
      <td>117</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2471</td>
      <td>2528</td>
      <td>+57</td>
      <td>57</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24883/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24883/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24883/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 50.76% is less than the suggested 90.0%
- `navier_stokes` new line coverage rate 67.24% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad committed:
Create NavierStokesProblem doc

lindsayad committed:
Update lid_driven_fps for Mass -> MassKernel

lindsayad committed:
Remove old-style input syntax

lindsayad committed:
Cleanup NavierStokesProblem code/input

- Provide more a more helpful parameter set and error messages
- Remove unnecessary methods

lindsayad committed:
Temporarily remove split size sum checking to allow nests

lindsayad committed:
Make NSFVPressurePin work for finite elements

Might need to rename this now

lindsayad committed:
Refine testing of the nested field split NS input

lindsayad committed:
Remove old syntax from FE lid driven tests spec

lindsayad committed:
More test refinement

lindsayad committed:
Give input files better names

lindsayad committed:
Try updating hybrid fsp inputs

Use user object for pressure pinning instead and remove the 'L'
matrix which is unused

However, this is exodiffing

lindsayad committed:
Restore hybrid test directory to pre-fsp work

lindsayad committed:
Sanity check for sizes of both top and nested splits

Also remove no longer needed `__FUNCT__` macros

lindsayad committed:
Add tests for nested splits

lindsayad committed:
Add name attribute to DM_Moose

lindsayad committed:
Get all boundary IDs when setting up the DM

This was a tricky distributed mesh failure to resolve. There were only
three local boundary IDs while there were four globally so the DM was
just taking the first three boundary names from the parser, and generally
that won't give you the local boundary IDs!

lindsayad committed:
Mass kernel documentation

lindsayad committed:
Add unside-by-var framework test

lindsayad committed:
Test Olshanskii commutation

lindsayad committed:
Fix and add test for pressure mass matrix preconditioning

lindsayad committed:
Destruct _L

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/1960576/) on 6bd9e52 : invalidated by @lindsayad

GiudGiud commented on modules/navier_stokes/include/problems/NavierStokesProblem.h:
```diff
@@ -0,0 +1,108 @@
+
+  /// This will end up being the same length as \p _schur_fs_index. Let's give an example of what
+  /// this data member means. If the user sets "schur_fs_index = '1'", then this means the Schur
+  /// complement field split is nested within another field split, and the Schr complement field
```
```suggestion
  /// complement field split is nested within another field split, and the Schur complement field
```

GiudGiud commented on modules/navier_stokes/src/problems/NavierStokesProblem.C:
```diff
@@ -0,0 +1,344 @@
+      "schur_fs_index",
+      {},
+      "if not provided then the top field split is assumed to be the "
+      "Schur split. This is a vector allow recursive nesting");
```
```suggestion
      "Schur split. This is a vector to allow recursive nesting");
```

GiudGiud commented on modules/navier_stokes/src/problems/NavierStokesProblem.C:
```diff
@@ -0,0 +1,344 @@
+}
+
+KSP
+NavierStokesProblem::findSchurKSP(KSP node, const unsigned int tree_position)
```
why not a reference on the  KSP argument?

lindsayad replied:
Every Petsc type is actually a pointer under the hood so we are only copying a pointer

GiudGiud commented on modules/navier_stokes/src/problems/NavierStokesProblem.C:
```diff
@@ -0,0 +1,344 @@
+
+  auto sub_ksp_index = *it;
+
+  ierr = KSPGetPC(node, &fs_pc);
+  LIBMESH_CHKERR2(this->comm(), ierr);
+  PetscObjectTypeCompare((PetscObject)fs_pc, PCFIELDSPLIT, &is_fs);
+  LIBMESH_CHKERR2(this->comm(), ierr);
+  if (!is_fs)
+    mooseError("Not a field split. Please check the 'schur_fs_index' parameter");
+  // Need to call this before getting the sub ksps
+  ierr = PCSetUp(fs_pc);
+  LIBMESH_CHKERR2(this->comm(), ierr);
+  ierr = PCFieldSplitGetSubKSP(fs_pc, &num_splits, &subksp);
+  LIBMESH_CHKERR2(this->comm(), ierr);
+  next_ksp = subksp[sub_ksp_index];
+
+  ierr = PCFieldSplitGetISByIndex(fs_pc, sub_ksp_index, &is);
+  LIBMESH_CHKERR2(this->comm(), ierr);
+
+  _index_sets.push_back(is);
+
+  ierr = PetscFree(subksp);
+  LIBMESH_CHKERR2(this->comm(), ierr);
```
needs comments

GiudGiud commented on modules/navier_stokes/src/problems/NavierStokesProblem.C:
```diff
@@ -0,0 +1,344 @@
+  }
+
+  ierr = PetscFree(subksp);
+  LIBMESH_CHKERR2(this->comm(), ierr);
```
could use comments throughout

probably should have a reviewer who s more versed in petsc calls. @fdkong @oanaoana ?

GiudGiud commented on modules/navier_stokes/src/problems/NavierStokesProblem.C:
```diff
@@ -0,0 +1,344 @@
+    global_Q =
+        static_cast<PetscMatrix<Number> &>(getNonlinearSystemBase(0).getMatrix(massMatrixTagID()))
+            .mat();
+  // The Poisson operator matrix
```
is this the Olshanki one? if so specify for velocity

GiudGiud commented on modules/navier_stokes/doc/content/source/problems/NavierStokesProblem.md:
```diff
@@ -0,0 +1,96 @@
+Least Squares Commutator (LSC) preconditioner, also known as the BFBt
+preconditioner, developed by Elman in [!citep](elman1999preconditioning) and
+[!citep](elman2006block).
+
```
add a link to the PETSc page on LSC?

GiudGiud commented on modules/navier_stokes/src/problems/NavierStokesProblem.C:
```diff
@@ -0,0 +1,344 @@
+    ierr = MatDestroy(&mat);
+    LIBMESH_CHKERR2(this->comm(), ierr);
+  }
+  for (auto & mat : intermediate_Ls)
```
why do we store them if we just end up destroy them here?

lindsayad replied:
It's just an easy way to progress to the sub-matrix representing the velocity-pressure dofs we are going to do LSC for. I could do more of a recursion that only ever leads to storage of one parent sub-matrix

GiudGiud commented on modules/navier_stokes/src/problems/NavierStokesProblem.C:
```diff
@@ -0,0 +1,344 @@
+  else
+  {
+    mooseAssert(our_parent_Q, "This should be non-null");
+    if (_commute_lsc || _pressure_mass_matrix_as_pre)
+    {
+      ierr =
+          MatCreateSubMatrix(our_parent_Q, pressure_is, pressure_is, MAT_REUSE_MATRIX, &_Q_scale);
+      LIBMESH_CHKERR2(this->comm(), ierr);
+    }
+    else
+    {
+      ierr =
+          MatCreateSubMatrix(our_parent_Q, velocity_is, velocity_is, MAT_REUSE_MATRIX, &_Q_scale);
+      LIBMESH_CHKERR2(this->comm(), ierr);
+    }
```
comment the else to distringuish which path (Olshanki pc ?)

lindsayad commented on framework/doc/content/source/kernels/MassKernel.md:
```diff
@@ -0,0 +1,14 @@
+# MassKernel
```
Maybe just `MassMatrix`?

lindsayad committed:
MassKernel -> MassMatrix

lindsayad committed:
FVMass -> FVMassMatrix

lindsayad committed:
Apply suggestions from code review

- doxygen
- fix doc string
- Add matrix tag parameter error checks in mass kernels
- Remove large PBS run script
- Fix 'oftype' typos
- Create validity checking function to follow DRY principles
- Delete correct data member in the MOOSE DM before allocating replacement
- Fix Schur typo in NavierStokesProblem
- Add a crap-ton of comments to NavierStokesProblem
- Add link to petsc LSC man page

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad commented on framework/src/splits/Split.C:
```diff
@@ -34,6 +34,12 @@ Split::validParams()
   params.addParam<std::vector<std::string>>(
       "splitting", {}, "The names of the splits (subsystems) in the decomposition of this split");
+  params.addParam<std::vector<BoundaryName>>(
+      "unside_by_var_boundary_name",
```
Dmitry was probably the original author of that name. But I like it fine. I think it conveys the same meaning as `remove_side` which conveys what the parameter is doing: it's removing the dofs on a side from the split

GiudGiud replied:
I unlike it but it s ok. I think when set this up in a Physics for NS we will access this parameter programmatically

lindsayad commented on test/tests/preconditioners/fsp/nested-split.i:
```diff
@@ -0,0 +1,117 @@
+      petsc_options_iname = '-ksp_type'
+      petsc_options_value = 'fgmres'
+    []
+      [u]
```
This is manual indenting to convey the different levels of the split. It's more clear to me than having all split levels have the same indentation

lindsayad commented on framework/src/utils/PetscDMMoose.C:
```diff
@@ -380,8 +456,33 @@ DMMooseSetUnSides(DM dm, const std::set<std::string> & unsides)
+  if (dm->setupcalled)
+    SETERRQ(PETSC_COMM_SELF, PETSC_ERR_ARG_WRONGSTATE, "Not for an already setup DM");
+  if (dmm->_sides)
+    delete dmm->_sides;
```
excellent catch. That's an embarrassing mistake

lindsayad commented on framework/src/utils/PetscDMMoose.C:
```diff
@@ -740,6 +831,24 @@ DMMooseGetEmbedding_Private(DM dm, IS * embedding)
+            if (var == v)
+              eligible_bids.insert(bid);
+
+          ConstBndNodeRange & bnodes = *dmm->_nl->mesh().getBoundaryNodeRange();
```
To my knowledge we don't have that. @roystgnr anything on the libmesh side? I perused the `BoundaryInfo` doxygen and didn't see anything but I could have overlooked something

roystgnr replied:
There isn't, and tragically it wouldn't currently be faster - internally we store boundary nodes in a `multimap<const Node *, boundary_id_type>`, so there's no way to iterate over only one id; the tree structure only sorts on `Node *` as its key.

I've been wanting to change that container for a while, and *really* wanting to change it during the days in which I was trying to hunt down a self-consistency failure in it last week, but something sweeping enough to allow independently fast iteration by both node and id would be a pretty major (and technically API breaking - somehow I let someone slip a raw `get_nodeset_map()` in there long ago...) upgrade.

lindsayad committed:
Add diagonal of A scaling test

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/2006263/) on b6c3e68 : invalidated by @lindsayad

GiudGiud commented on modules/navier_stokes/doc/content/source/problems/NavierStokesProblem.md:
```diff
@@ -0,0 +1,97 @@
+proposed by Elman approximates the Schur complement inverse via
+
+\begin{equation}
+\hat{S}^{-1} := (B\hat{M}_u^{-1}B^T)^{-1}B\hat{M}_u^{-1}A\hat{M}_u^{-1}B_T(B\hat{M}_u^{-1}B_T)^{-1}
```
if you could write M in the context of the A, B, B matrix above, I think it would make things clearer

it s some sort of preconditioning applied on B here imo

lindsayad replied:
The next line explains what M is:

> where $\hat{M}_u$ is a diagonal approximation of the velocity mass matrix

It doesn't have anything to do with A or B

> it s some sort of preconditioning applied on B here imo

What is this sentence referring to?

GiudGiud replied:
It s referring to adding M in the expression

In the petsc manual for LSC, there is no M in the definition of S-1

lindsayad replied:
That's a PETSc issue. I didn't update that documentation in https://gitlab.com/petsc/petsc/-/merge_requests/6642

GiudGiud commented on modules/navier_stokes/doc/content/source/problems/NavierStokesProblem.md:
```diff
@@ -0,0 +1,97 @@
+[!param](/Problem/NavierStokesProblem/mass_matrix) is not provided, then
+$\hat{M}_u$ will be populated with the diagonal of $A$. If
+[!param](/Problem/NavierStokesProblem/L_matrix) is not provided, then it will be
+computed automatically via
```
from your litterature study, can you make recommendations on these parameters based on the flow regime?

alternatively, can you point the references you used to the selection of each parameter?

For example for commute_lsc,I know to go read the Olshankii paper to under why I would use that

lindsayad replied:
I've added a section at the bottom of this file offering guidance on preconditioner selection

GiudGiud commented on modules/navier_stokes/include/userobjects/NSFVPressurePin.h:
```diff
@@ -28,17 +28,17 @@ class NSFVPressurePin : public GeneralUserObject,
 
   /// The thread 0 copy of the pressure variable
-  INSFVPressureVariable * const _p;
+  const MooseVariableFieldBase & _p;
```
look like my UO should be NSPressurePin now?

GiudGiud reviewed (commented):
thanks for the comments. It s a lot clearer to me

lindsayad committed:
Add vector implementation elman test

Allows for quick comparison of olshanskii and elman preconditioners

lindsayad committed:
Make recommendations to users about what preconditioners to use

fdkong commented on modules/navier_stokes/doc/content/source/problems/NavierStokesProblem.md:
```diff
@@ -88,6 +88,66 @@ A final option available to users is the [!param](/Problem/NavierStokesProblem/u
+paper is that Olshanskii is better performing, although the largest kinematic
+viscosity explored in that article direclty comparing the two methods is
+$10^{-1}$. The Olshanskii preconditioner is known to be sub-optimal for high
+viscosities. In the limit of Stokes flow in which the advective term is dropped,
```
What do you mean by "high viscosities"? Slow fluid?

lindsayad replied:
It doesn't necessarily have to be slow, but it tends to be for high viscosity (like maple syrup)

lindsayad replied:
in this test input, the peak velocity is the same for all cases

fdkong commented on modules/navier_stokes/doc/content/source/problems/NavierStokesProblem.md:
```diff
@@ -88,6 +88,66 @@ A final option available to users is the [!param](/Problem/NavierStokesProblem/u
+| 1         | 11     | 12     | 12     |
+| $10^{-2}$ | 17     | 18     | 17     |
+
+Pressure
```
"Pressure" indicates that you are using mass matrix?

lindsayad replied:
yes, pressure mass matrix

fdkong commented on modules/navier_stokes/doc/content/source/problems/NavierStokesProblem.md:
```diff
@@ -88,6 +88,66 @@ A final option available to users is the [!param](/Problem/NavierStokesProblem/u
+`steady_vector_fsp.i`, and `steady_vector_fsp_stokes.i` inputs respectively. To
+focus the comparison on the preconditioning of the Schur complement itself, all
+multigrid preconditioners were replaced with LU decompositions (`-pc_type lu
+-pc_factor_mat_solver_type mumps`). In the `steady_vector_fsp_stokes.i` input,
```
I want to see the result of "-snes_view" where I will have every single detail of your setup.

lindsayad replied:
<details>
<summary>Example SNES view</summary>
```
SNES Object: 1 MPI process
  type: newtonls
  maximum iterations=50, maximum function evaluations=10000
  tolerances: relative=1e-08, absolute=1e-50, solution=0.
  total number of linear solver iterations=8
  total number of function evaluations=4
  norm schedule ALWAYS
  SNESLineSearch Object: 1 MPI process
    type: bt
      interpolation: cubic
      alpha=1.000000e-04
    maxstep=1.000000e+08, minlambda=1.000000e-12
    tolerances: relative=1.000000e-08, absolute=1.000000e-15, lambda=1.000000e-08
    maximum iterations=40
  KSP Object: 1 MPI process
    type: fgmres
      restart=300, using Classical (unmodified) Gram-Schmidt Orthogonalization with no iterative refinement
      happy breakdown tolerance 1e-30
    maximum iterations=10000, initial guess is zero
    tolerances:  relative=0.0001, absolute=1e-50, divergence=1e+100
    right preconditioning
    using UNPRECONDITIONED norm type for convergence test
  PC Object: 1 MPI process
    type: fieldsplit
      FieldSplit with Schur preconditioner, factorization FULL
      Preconditioner for the Schur complement formed from S itself
      Split info:
      Split number 0 Defined by IS
      Split number 1 Defined by IS
      KSP solver for A00 block
        KSP Object: (fieldsplit_u_) 1 MPI process
          type: gmres
            restart=300, using Classical (unmodified) Gram-Schmidt Orthogonalization with no iterative refinement
            happy breakdown tolerance 1e-30
          maximum iterations=10000, initial guess is zero
          tolerances:  relative=0.01, absolute=1e-50, divergence=10000.
          right preconditioning
          using UNPRECONDITIONED norm type for convergence test
        PC Object: (fieldsplit_u_) 1 MPI process
          type: hypre
            HYPRE BoomerAMG preconditioning
              Cycle type V
              Maximum number of levels 25
              Maximum number of iterations PER hypre call 1
              Convergence tolerance PER hypre call 0.
              Threshold for strong coupling 0.25
              Interpolation truncation factor 0.
              Interpolation: max elements per row 0
              Number of levels of aggressive coarsening 0
              Number of paths for aggressive coarsening 1
              Maximum row sums 0.9
              Sweeps down         1
              Sweeps up           1
              Sweeps on coarse    1
              Relax down          symmetric-SOR/Jacobi
              Relax up            symmetric-SOR/Jacobi
              Relax on coarse     Gaussian-elimination
              Relax weight  (all)      1.
              Outer relax weight (all) 1.
              Maximum size of coarsest grid 9
              Minimum size of coarsest grid 1
              Using CF-relaxation
              Not using more complex smoothers.
              Measure type        local
              Coarsen type        Falgout
              Interpolation type  classical
              SpGEMM type         hypre
          linear system matrix = precond matrix:
          Mat Object: (fieldsplit_u_) 1 MPI process
            type: seqaij
            rows=578, cols=578
            total: nonzeros=16900, allocated nonzeros=16900
            total number of mallocs used during MatSetValues calls=0
              using I-node routines: found 266 nodes, limit used is 5
      KSP solver for S = A11 - A10 inv(A00) A01
        KSP Object: (fieldsplit_p_) 1 MPI process
          type: fgmres
            restart=300, using Classical (unmodified) Gram-Schmidt Orthogonalization with no iterative refinement
            happy breakdown tolerance 1e-30
          maximum iterations=10000, initial guess is zero
          tolerances:  relative=0.01, absolute=1e-50, divergence=10000.
          right preconditioning
          using UNPRECONDITIONED norm type for convergence test
        PC Object: (fieldsplit_p_) 1 MPI process
          type: lsc
            KSP Object: (fieldsplit_p_lsc_) 1 MPI process
              type: fgmres
                restart=300, using Classical (unmodified) Gram-Schmidt Orthogonalization with no iterative refinement
                happy breakdown tolerance 1e-30
              maximum iterations=10000, initial guess is zero
              tolerances:  relative=0.1, absolute=1e-50, divergence=10000.
              right preconditioning
              using UNPRECONDITIONED norm type for convergence test
            PC Object: (fieldsplit_p_lsc_) 1 MPI process
              type: hypre
                HYPRE BoomerAMG preconditioning
                  Cycle type V
                  Maximum number of levels 25
                  Maximum number of iterations PER hypre call 1
                  Convergence tolerance PER hypre call 0.
                  Threshold for strong coupling 0.25
                  Interpolation truncation factor 0.
                  Interpolation: max elements per row 0
                  Number of levels of aggressive coarsening 0
                  Number of paths for aggressive coarsening 1
                  Maximum row sums 0.9
                  Sweeps down         1
                  Sweeps up           1
                  Sweeps on coarse    1
                  Relax down          symmetric-SOR/Jacobi
                  Relax up            symmetric-SOR/Jacobi
                  Relax on coarse     Gaussian-elimination
                  Relax weight  (all)      1.
                  Outer relax weight (all) 1.
                  Maximum size of coarsest grid 9
                  Minimum size of coarsest grid 1
                  Using CF-relaxation
                  Not using more complex smoothers.
                  Measure type        local
                  Coarsen type        Falgout
                  Interpolation type  classical
                  SpGEMM type         hypre
              linear system matrix = precond matrix:
              Mat Object: 1 MPI process
                type: seqaij
                rows=578, cols=578
                total: nonzeros=16900, allocated nonzeros=16900
                total number of mallocs used during MatSetValues calls=0
                  using I-node routines: found 266 nodes, limit used is 5
            KSP Object: (fieldsplit_p_lsc_mass_) 1 MPI process
              type: gmres
                restart=30, using Classical (unmodified) Gram-Schmidt Orthogonalization with no iterative refinement
                happy breakdown tolerance 1e-30
              maximum iterations=10000, initial guess is zero
              tolerances:  relative=0.1, absolute=1e-50, divergence=10000.
              left preconditioning
              using PRECONDITIONED norm type for convergence test
            PC Object: (fieldsplit_p_lsc_mass_) 1 MPI process
              type: hypre
                HYPRE BoomerAMG preconditioning
                  Cycle type V
                  Maximum number of levels 25
                  Maximum number of iterations PER hypre call 1
                  Convergence tolerance PER hypre call 0.
                  Threshold for strong coupling 0.25
                  Interpolation truncation factor 0.
                  Interpolation: max elements per row 0
                  Number of levels of aggressive coarsening 0
                  Number of paths for aggressive coarsening 1
                  Maximum row sums 0.9
                  Sweeps down         1
                  Sweeps up           1
                  Sweeps on coarse    1
                  Relax down          symmetric-SOR/Jacobi
                  Relax up            symmetric-SOR/Jacobi
                  Relax on coarse     Gaussian-elimination
                  Relax weight  (all)      1.
                  Outer relax weight (all) 1.
                  Maximum size of coarsest grid 9
                  Minimum size of coarsest grid 1
                  Using CF-relaxation
                  Not using more complex smoothers.
                  Measure type        local
                  Coarsen type        Falgout
                  Interpolation type  classical
                  SpGEMM type         hypre
              linear system matrix = precond matrix:
              Mat Object: 1 MPI process
                type: seqaij
                rows=81, cols=81
                total: nonzeros=625, allocated nonzeros=625
                total number of mallocs used during MatSetValues calls=0
                  not using I-node routines
          linear system matrix = precond matrix:
          Mat Object: (fieldsplit_p_) 1 MPI process
            type: schurcomplement
            rows=81, cols=81
              Schur complement A11 - A10 inv(A00) A01
              A11
                Mat Object: (fieldsplit_p_) 1 MPI process
                  type: seqaij
                  rows=81, cols=81
                  total: nonzeros=81, allocated nonzeros=81
                  total number of mallocs used during MatSetValues calls=0
                    not using I-node routines
              A10
                Mat Object: 1 MPI process
                  type: seqaij
                  rows=81, cols=578
                  total: nonzeros=3362, allocated nonzeros=3362
                  total number of mallocs used during MatSetValues calls=0
                    not using I-node routines
              KSP solver for A00 block viewable with the additional option -fieldsplit_u_ksp_view
              A01
                Mat Object: 1 MPI process
                  type: seqaij
                  rows=578, cols=81
                  total: nonzeros=3362, allocated nonzeros=3362
                  total number of mallocs used during MatSetValues calls=0
                    using I-node routines: found 266 nodes, limit used is 5
    linear system matrix = precond matrix:
    Mat Object: () 1 MPI process
      type: seqaij
      rows=659, cols=659
      total: nonzeros=23705, allocated nonzeros=24249
      total number of mallocs used during MatSetValues calls=0
        using I-node routines: found 347 nodes, limit used is 5
```
</details>

fdkong commented on modules/navier_stokes/doc/content/source/problems/NavierStokesProblem.md:
```diff
@@ -88,6 +88,66 @@ A final option available to users is the [!param](/Problem/NavierStokesProblem/u
+| 100       | 6      | 6      | 6      |
+| 1         | 6      | 6      | 6      |
+| $10^{-2}$ | 21     | 21     | 21     |
+
```
Super cool  comparison and data.   Why `Elman` is so sensitive to grid size? I assume "n" is the grid size.

lindsayad replied:
It's insensitive to grid size for viscous flows, but it is known to scale ("mildly" is oft-said in the literature, but can be more severe than "mild") with grid size for advection dominated flows. The *Scalable block preconditioners for linearized Navier-Stokes equations at high Reynolds number* reference shows some nice eigenvalue plots for Elman vs. Olshanskii

fdkong reviewed (commented):
Really cool stuff. I want to see more detail of your solver setup

lindsayad committed:
NSFVPressurePin -> NSPressurePin

GiudGiud commented on modules/navier_stokes/doc/content/source/problems/NavierStokesProblem.md:
```diff
@@ -88,6 +88,68 @@ A final option available to users is the [!param](/Problem/NavierStokesProblem/u
+focus the comparison on the preconditioning of the Schur complement itself, all
+multigrid preconditioners were replaced with LU decompositions (`-pc_type lu
+-pc_factor_mat_solver_type mumps`). In the `steady_vector_fsp_stokes.i` input,
+an `INSFVMomentumAdvection` block was added as well.
```
an [INSFVMomentumAdvection.md] kernel

lindsayad committed:
Documentation fixes

- Reference `INSADMomentumAdvection` markdown file
- Remove duplicate issues
- NSFVPressurePin -> NSPressurePin in the newsletter

lindsayad committed:
Add a couple more code comments in NavierStokesProblem

lindsayad commented:
thank you for the review!

loganharbour commented:
This wrekt petsc alt: https://civet.inl.gov/event/164129/

lindsayad commented:
get wrekt
